src/cuda_sampling.cu

```diff
@@ -0,0 +1,605 @@
+  LaunchPopulateIndices(indices_in.data(), vocab_size, batch_size, stream);
+  size_t temp_storage_bytes = 0;
+  GetTempStorageSize<float>(scores.data(), indices_in.data(), offsets_gpu.data(), vocab_size * batch_size,
+                            batch_size, stream, /*is_descending*/true, temp_storage_bytes);
+  auto temp_buffer = CudaMallocArray<float>(temp_storage_bytes / sizeof(float));
+  std::span<float> temp_span{temp_buffer.get(), temp_storage_bytes / sizeof(float)};
+  LaunchSortPairs<float>(temp_span.data(), temp_storage_bytes, scores.data(), scores_out,
+                         indices_in.data(), indices_out, vocab_size * batch_size, batch_size, offsets_gpu.data(),
+                         stream, /*is_descending*/true);
+}
+
+void LaunchGetTopKSubsetFullSort(cudaStream_t stream, float* scores_in, float* scores_out, int* indices_out, int vocab_size, int batch_size, int k) {
+  // Softmax and sort scores
+  auto scores_buffer = CudaMallocArray<float>(vocab_size * batch_size);
+  auto indices_buffer = CudaMallocArray<int>(vocab_size * batch_size);
+  std::span<float> scores_sorted{scores_buffer.get(), static_cast<size_t>(vocab_size * batch_size)};
+  std::span<int> indices_sorted{indices_buffer.get(), static_cast<size_t>(vocab_size * batch_size)};
+  SoftmaxAndSort(stream, scores_in, scores_sorted.data(), indices_sorted.data(), vocab_size, batch_size, 1.0);
+  // Get top k subset
+  dim3 grid(batch_size * int(1 + k / 256), 1, 1);
+  dim3 block(256, 1, 1);
+  GetTopKSortedInputKernel<<<grid, block, 0, stream>>>(indices_sorted.data(), indices_out, scores_sorted.data(), scores_out, batch_size, vocab_size, k);
+}
+
+void GetTopKSubset(cudaStream_t stream, float* scores_in, float* scores_out, int* indices_out, int vocab_size, int batch_size, int k, float temperature) {
+  // Softmax scores
+  auto scores_buffer = CudaMallocArray<float>(vocab_size * batch_size);
+  std::span<float> scores_softmaxed{scores_buffer.get(), static_cast<size_t>(vocab_size * batch_size)};
+  DispatchBlockwiseSoftmaxForward<false>(&stream, scores_softmaxed.data(), const_cast<const float*>(scores_in), vocab_size, vocab_size, vocab_size, batch_size, temperature);
+  // Get top k subset
+  #define GetTopK(max_k)                       \
+  LaunchGetTopKSubset<max_k>(stream,                  \
+                             scores_softmaxed.data(), \
+                             scores_out,       \
+                             indices_out,      \
+                             vocab_size,       \
+                             batch_size,       \
+                             k)
+
+  if (k <= 4) {
+    GetTopK(4);
+  } else if (k <= 8) {
+    GetTopK(8);
+  } else if (k <= 16) {
+    GetTopK(16);
+  } else if (k <= 32) {
+    GetTopK(32);
+  } else if (k <= 64) {
+    GetTopK(64);
+  } else {
+    LaunchGetTopKSubsetFullSort(stream, scores_softmaxed.data(), scores_out, indices_out, vocab_size, batch_size, k);
```

Comment:
>scores_softmaxed [](http://example.com/codeflow?start=40&length=16)

scores_softmaxed is already SoftMax. SoftmaxAndSort invoked in LaunchGetTopKSubsetFullSort will do softmax again.